Added support for 'Tamron SP 15-30mm f/2.8 Di VC USD A012' and 'Tamron SP 90mm f/2.8 Di VC USD MACRO1:1'

mfg92 opened the pull request:
Added support for 'Tamron SP 15-30mm f/2.8 Di VC USD A012' and 'Tamron SP 90mm f/2.8 Di VC USD MACRO1:1'. But there is no distinction between the F004 and the F017 Versions of the 90mm.

Fixed Bug #1306 and #1304 (partly)

Test results:

$ ./bin/exiv2 -pa ../TamronSP15-30mmF2.8DiVCUSDA012.JPG | grep -ai lens
Exif.CanonCs.LensType                        Short       1  Tamron SP 15-30mm f/2.8 Di VC USD A012
Exif.CanonCs.Lens                            Short       3  15.0 - 30.0 mm
Exif.Canon.LensModel                         Ascii      74  TAMRON SP 15-30mm F/2.8 Di VC USD A012
Exif.Photo.LensSpecification                 Rational    4  15/1 30/1 0/1 0/1
Exif.Photo.LensModel                         Ascii      39  TAMRON SP 15-30mm F/2.8 Di VC USD A012
Exif.Photo.LensSerialNumber                  Ascii      11  0000000000

$ ./bin/exiv2 -pa ../TamronSP90mmF2.8DiVCUSDMacroF004.JPG | grep -ai lens
Exif.CanonCs.LensType                        Short       1  Tamron SP 90mm f/2.8 Di VC USD MACRO1:1
Exif.CanonCs.Lens                            Short       3  90.0 mm
Exif.Canon.LensModel                         Ascii      70  TAMRON SP 90mm F/2.8 Di VC USD MACRO1:1 F004

$ ./bin/exiv2 -pa ../TamronSP90mmF2.8DiVCUSDMacroF017.JPG | grep -ai lens
Exif.CanonCs.LensType                        Short       1  Tamron SP 90mm f/2.8 Di VC USD MACRO1:1
Exif.CanonCs.Lens                            Short       3  90.0 mm
Exif.Canon.LensModel                         Ascii      70  TAMRON SP 90mm F/2.8 Di VC USD MACRO1:1 F017
Exif.Photo.LensSpecification                 Rational    4  90/1 90/1 0/0 0/0
Exif.Photo.LensModel                         Ascii      70  TAMRON SP 90mm F/2.8 Di VC USD MACRO1:1 F017
Exif.Photo.LensSerialNumber                  Ascii      12  0000000000

clanmills commented:
mfg92:  Thank You for you contribution.  We have been very busy in July, August and September with a rewrite of our CMake support.  So, regrettably, your contribution is heavily conflicted.  May I ask you to redo your Pull Request relative to the 'master' and I'll integrate your changes this week.

mfg92 committed:
Added support for 'Tamron SP 15-30mm f/2.8 Di VC USD A012' and 'Tamron SP 90mm f/2.8 Di VC USD MACRO1:1'.

But there is no distinction between the F004 and the F017 Versions of the 90mm.
Fixed Bug #1306 and #1304 (partly)

mfg92 commented:
I have updated my fork and did the changes again (manually, beacuse git confused me).

clanmills commented:
Thank you very much for your contribution and please accept my apologies that I didn't service this immediately.  If you have any test images in which those lenses are being used, I'd like to extract the metadata and add it to our test suite.

We only adopted git in April this year when we released v0.26.  I'm still confused and baffled by git.  One day the fog will disappear for us both!

clanmills reviewed (commented):
I didn't write this code.  When the ID is unique, the lens will be correctly identified.  However, in this case, 137 is overloaded.  I like to have files in our test suite to be sure that the fix doesn't regress when additional 137 lens are added.  I'll take care of the test suite if you attach an image used by the lens.  Additionally, I only store the metadata in the test suite to save storage and to retain your privacy.

mfg92 commented:
I have to thank you for maintaining exiv2 :)
Only one of my test photos was taken by me:

TamronSP90mmF2.8DiVCUSDMacroF017.JPG
![tamronsp90mmf2 8divcusdmacrof017](https://user-images.githubusercontent.com/11365250/30716273-be2f4fb2-9f19-11e7-96c5-f670fd5849cd.JPG)

For images taken with the other lenses (TamronSP90mmF2.8DiVCUSDMacroF004, TamronSP15-30mmF2.8DiVCUSDA012, CanonEF100mmF2.8LMacroISUSM) i have searched the internet, so they are not from me. That is why i dont want to upload them here. But i can extract the metadata for you. What command should i use to extract them?

clanmills commented:
Thanks very much.  I've extracted the metadata with the command <code>$ exiv2 -ea --force --verbose ~/Desktop/30716273-be2f4fb2-9f19-11e7-96c5-f670fd5849cd.JPG</code>

I added your metadata as test/data/exiv2-g20.exv and updated test/bugfixes.sh to read your file.  Job Done.  Thanks for your contribution.

The changes are subtle.  With v0.26:
```
Exif.CanonCs.LensType                        Short       1  Canon EF 100mm f/2.8L Macro IS USM
```
With your change:
```
Exif.CanonCs.LensType                        Short       1  Tamron SP 90mm f/2.8 Di VC USD Macro 1:1
```

mfg92 commented:
I have extracted the metadata of the other images/lenses too, you can use them if you like.

[testPhotos.zip](https://github.com/Exiv2/exiv2/files/1322926/testPhotos.zip)

`[matze@matze-antergos exiv2]$ ./build/bin/exiv2 -ea --force --verbose TamronSP*                                                                        
File 1/3: TamronSP15-30mmF2.8DiVCUSDA012.JPG                                                                                                          
Writing Exif data from TamronSP15-30mmF2.8DiVCUSDA012.JPG to ./TamronSP15-30mmF2.8DiVCUSDA012.exv                                                     
Writing XMP data from TamronSP15-30mmF2.8DiVCUSDA012.JPG to ./TamronSP15-30mmF2.8DiVCUSDA012.exv                                                      
File 2/3: TamronSP90mmF2.8DiVCUSDMacroF004.JPG                                                                                                        
Writing Exif data from TamronSP90mmF2.8DiVCUSDMacroF004.JPG to ./TamronSP90mmF2.8DiVCUSDMacroF004.exv                                                 
File 3/3: TamronSP90mmF2.8DiVCUSDMacroF017.JPG                                                                                                        
Writing Exif data from TamronSP90mmF2.8DiVCUSDMacroF017.JPG to ./TamronSP90mmF2.8DiVCUSDMacroF017.exv                                                 
Writing XMP data from TamronSP90mmF2.8DiVCUSDMacroF017.JPG to ./TamronSP90mmF2.8DiVCUSDMacroF017.exv
[matze@matze-antergos exiv2]$ ./build/bin/exiv2 -ea --force --verbose CanonEF100mmF2.8LMacroISUSM.jpg 
File 1/1: CanonEF100mmF2.8LMacroISUSM.jpg
Writing Exif data from CanonEF100mmF2.8LMacroISUSM.jpg to ./CanonEF100mmF2.8LMacroISUSM.exv
`

clanmills commented:
Very good.  Thanks.  Submitted: 3755d7c1
```
exiv2/0.26 $ exiv2 -pa --grep lenstype/i ~/Downloads/testPhotos/*.exv
/Users/rmills/Downloads/testPhotos/CanonEF100mmF2.8LMacroISUSM.exv  Exif.CanonCs.LensType                        Short       1  Canon EF 100mm f/2.8L Macro IS USM
/Users/rmills/Downloads/testPhotos/TamronSP15-30mmF2.8DiVCUSDA012.exv  Exif.CanonCs.LensType                        Short       1  Canon EF 300mm f/2.8L
/Users/rmills/Downloads/testPhotos/TamronSP90mmF2.8DiVCUSDMacroF004.exv  Exif.CanonCs.LensType                        Short       1  Canon EF 100mm f/2.8L Macro IS USM
/Users/rmills/Downloads/testPhotos/TamronSP90mmF2.8DiVCUSDMacroF017.exv  Exif.CanonCs.LensType                        Short       1  Canon EF 100mm f/2.8L Macro IS USM
exiv2/0.26 $ cd ~/gnu/github/exiv2/exiv2/
github/exiv2/exiv2 $ makes
make
      8      76     718
sudo make install
exiv2/exiv2 $ exiv2 -pa --grep lenstype/i ~/Downloads/testPhotos/*.exv
/Users/rmills/Downloads/testPhotos/CanonEF100mmF2.8LMacroISUSM.exv  Exif.CanonCs.LensType                        Short       1  Canon EF 100mm f/2.8L Macro IS USM
/Users/rmills/Downloads/testPhotos/TamronSP15-30mmF2.8DiVCUSDA012.exv  Exif.CanonCs.LensType                        Short       1  Tamron SP 15-30mm f/2.8 Di VC USD A012
/Users/rmills/Downloads/testPhotos/TamronSP90mmF2.8DiVCUSDMacroF004.exv  Exif.CanonCs.LensType                        Short       1  Tamron SP 90mm f/2.8 Di VC USD Macro 1:1
/Users/rmills/Downloads/testPhotos/TamronSP90mmF2.8DiVCUSDMacroF017.exv  Exif.CanonCs.LensType                        Short       1  Tamron SP 90mm f/2.8 Di VC USD Macro 1:1
561 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2 $ 
```
